overlord/servicestate/quota_control.go

```diff
@@ -266,6 +292,38 @@ func UpdateQuota(st *state.State, name string, updateOpts QuotaGroupUpdate) (*st
 	return ts, nil
 }
 
+// remove a string item at index i from the string slice,
+// it maintains the ordering of the original slice.
+func remove(slice []string, i int) []string {
+	return append(slice[:i], slice[i+1:]...)
+}
+
+// ensureSnapServicesAbsentFromSubGroups removes all service references of a snap in
+// sub-groups related to the group of the snap, and returns the groups that were modified.
+func ensureSnapServicesAbsentFromSubGroups(grp *quota.Group, snap string, allGrps map[string]*quota.Group) ([]*quota.Group, error) {
+	// We can assume that when removing a snap, that if it has services in
+	// sub-groups that there will be only one level of sub-groups, so we can
+	// avoid checking further sub-groups.
+	var modifiedGrps []*quota.Group
+	for _, name := range grp.SubGroups {
+		subgrp, ok := allGrps[name]
+		if !ok {
+			return nil, fmt.Errorf("non-existent sub-group %q", name)
+		}
+
+		for idx, svc := range subgrp.Services {
+			parts := strings.Split(svc, ".")
+			if parts[0] == snap {
+				// found a service that matches the snap we are removing,
+				// so remove that too
+				subgrp.Services = remove(subgrp.Services, idx)
+				modifiedGrps = append(modifiedGrps, subgrp)
+			}
+		}
+	}
```

Comment:
The first comment in the function said that if we found services, we should not need to check furtehr subgroups. But this function is never doing the latter. Shouldn't we have here a loop where we recurse into subgroups, if no services were found?

Reply:
mmm, but here we don't know whether this is a group with snaps. My point was: if we didn't find any services, then it's not a mixed group, and therefore we should recurse into its children, but we are not doing that here. At least, this is what I understand from the comment at the top of this function.

Reply:
We know the group has snap, because the function is ever only called for a group where we found the snap we are trying to remove. So when this function is called, it's always for a group containing snaps. Either it contains sub-groups (any number) as well, in which case there can only be one level of sub-groups (enforced during group creation/update), or it doesn't contain any sub-groups.

Reply:
Ah, I see! Then may I suggest changing the comment at the beginning of this function from

	// When removing a snap we assume that if it has services in
	// sub-groups there will be only one level of sub-groups, so we can
	// avoid checking any nested sub-groups.

to something like

> If a snap has services in sub-groups, the services must be in the first level of sub-groups; that's why the code here does not check for nested sub-groups.

Reply:
Thank you for the feedback! Will change the description!